### DISCUSSION TITLE
Salesforce and recurring donation problem with custom fields

### DISCUSSION THREAD

**Discussion by diegoinidis** [impress-org/givewp#6583] [Developer Q&A] [answered]
Hi, I've been referred from Priority Support to write my problem here.
I'm working to connect several donations forms with Salesforce with some custom fields. I'm using Give with add-ons Recurring Donations and Salesforce.
I'm sending custom fields data to Salesforce through give_salesforce_donation_args I'm not having any problem with one-time donations however when I want to create a recurrent donation, the method `SyncRecurringDonationController::createRecurringDonationAndInitialDonation` check if donation renewal exists and then delete the opportunity record created before so than the custom fields data is lost.
I could try to send this data to Salesforce through the contact or even recurring donations repositories, however I have no filters to add new information.
I can see on logs a sequence like this in any recurring donation:

  ```
Create Salesforce record: contact
  "data": {
          "FirstName": "ALBERTO",
          "LastName": "DEL CARDO",
          "Email": "info04@loquesea.es"
      },
      "body": {
          "Object Class": "stdClass",
          "id": "003Fg000000f1UnIAI",
          "success": true,
          "errors": []
      },
  Create Salesforce record: opportunity
  "data": {
          "Type": "Donation",
          "Name": "Donation #422",
          "ContactId": "003Fg000000f1UnIAI",
          "Amount": "50.000000",
          "AccountId": "001Fg000000uDx3IAE",
          "CloseDate": "2022-10-18",
          "StageName": "Pledged",
          "DEBRA_SendDonationCertificate__c": true,
          "DEB_Codigo_postal__c": "00000",
          "DEB_Provincia__c": "pataplin",
          "DEB_Metodo_de_pago__c": "PAYPAL",
          "DEB_Tipo_de_documento__c": "DNI",
          "DEB_Numero_de_documento__c": "10000000-Z",
          "DEB_Idioma__c": "Español",
          "DEB_Origen_donacion__c": "WEB"
      },
      "body": {
          "Object Class": "stdClass",
          "id": "006Fg000000hSp3IAE",
          "success": true,
          "errors": []
      },
  Create Salesforce record: npe03__Recurring_Donation__c
  "data": {
          "Name": "Recurring Donation #30",
          "npe03__Contact__c": "003Fg000000f1UnIAI",
          "npe03__Amount__c": "50.0000000000",
          "npe03__Installment_Period__c": "Monthly",
          "npsp__InstallmentFrequency__c": "1",
          "npsp__RecurringType__c": "Open",
          "npe03__Open_Ended_Status__c": "Open",
          "npe03__Installments__c": "0",
          "npe03__Date_Established__c": "2022-10-18",
          "npsp__Day_of_Month__c": "18"
      },
      "body": {
          "Object Class": "stdClass",
          "id": "a09Fg0000003Lj3IAE",
          "success": true,
          "errors": []
      },
  
  Deleted Salesforce record: opportunity/006Fg000000hSp3IAE
  
  Updated Salesforce record: opportunity/006Fg000000hSqfIAE
  "data": {
          "npe03__Recurring_Donation__c": "a09Fg0000003Lj3IAE",
          "Name": "Donation #422 (Recurring Installment)",
          "StageName": "Pledged"
      },
      "body": null,
  
  Updated Salesforce record: opportunity/006Fg000000hSqfIAE
  "data": {
          "StageName": "Closed Won"
      },
      "body": null,

```
How could I process these custom data without altering the addon source code?
Thanks!
Diego Mateos

**Comment by kjohnson** [comment]
Hey @diegoinidis, I'm not sure if this is a bug on the `SyncRecurringDonationController` class or if we should add additional filters to compensate. I'll need to consult with @jonwaldstein. He is not in the office this week, but I'll follow-up with him when he returns.

**Comment by jonwaldstein** [comment, marked as answer]
@diegoinidis Thanks for reaching out.  

(thinking out load)

If I remember correctly, some versions of Salesforce will automatically create "pledged" donations based on the parameters of the recurring donation record (end close date / installments).  Looking back through our codebase - we are swapping the initial donation created with the _pledged donation_ to prevent duplicate records.  I think what you have discovered is the initial donation that holds the filter with custom fields is the one being swapped out for the auto-generated donation created by salesforce.  This is obviously problematic when it comes to custom data.

Ideally, the swapping of the initial donation & the pledged donation would happen in reverse order so the initial data is preserved.  I would have to look back into the Salesforce api to see if that's possible.  If not, then we will need to add another filter or two to compensate for the nuances of Salesforce recurring donations & auto-generated pledged donations.

**As a work-around**, if you were able to setup your Salesforce account _without_ auto-generated pledged donation within your recurring donations - this would not be a problem because the initial donation would never be deleted.

**Comment by jason-best** [comment]
@jonwaldstein I am having a similar problem and turning off "auto-create" in Salesforce is causing other issues. Here is what is happening:

I am passing custom metadata to the salesforce integration. It works wonderfully for individual donations, but there are issues with recurring donations.

If "Installment Opportunity Auto-Creation" is turned OFF in SF, the recurring record is never created, but the custom metadata is passed as expected.

If "Installment Opportunity Auto-Creation" is turned ON in SF, the recurring record is created, but the custom metadata is not passed.

Is this a known issue and are you aware of any resolutions?

**Comment by jason-best** [comment]
@jonwaldstein I have added a new filter to UpdateSalesforceOpportunityAsRecurringInstallmentAction.php on line 79. 
`$filteredArgs = apply_filters('give_salesforce_donation_update_args', [], $donationId);
		$args=array_merge($args,$filteredArgs);`
I am then able to add my custom metadata to the updated Opportunity. It is working well. 

How could I officially contribute this to the plugin?

**Reply by jonwaldstein** [reply]
@jason-best thanks for the update, glad to hear the custom filter is working.  Since this add-on is a private repo the best way to contribute would be to add a feedback item to our canny board https://feedback.givewp.com/ and provide your solution.  This way folks can vote on it and our company can decide to move this work into our queue.

**Reply by jonwaldstein** [reply]
Hey @jason-best!

I'm currently making an update to Salesforce and wanted to circle back here.  Are you still using the filter directly in the codebase?

**Reply by jason-best** [reply]
@jonwaldstein Yes. We currently are doing the filter in the codebase and just have to copy over on updates.

**Reply by jonwaldstein** [reply]
@jason-best okay thanks!  are you using that exact snippet you shared?  I want to make sure your filter does not break when we release this.  Would you prefer option 1 or 2 below?

**Option 1: filter directly on base args**
<img width="681" alt="Screenshot 2025-06-23 at 9 10 11 AM" src="https://github.com/user-attachments/assets/507f3e7e-4986-4788-a5b3-86cdae560986" />

**Option 2: filter merged with base args**
<img width="834" alt="Screenshot 2025-06-23 at 9 11 28 AM" src="https://github.com/user-attachments/assets/7419302a-59f8-4f38-8a53-24542df04ac9" />

**Reply by jonwaldstein** [reply]
Hey @jason-best circling back here to let you know we release Salesforce 1.2.0 with the following new filters:

- `give_salesforce_donation_update_args`
- `give_salesforce_recurring_donation_create_args`
- `give_salesforce_contact_create_args`
- `give_salesforce_contact_search_by_full_name`

**Comment by merguk** [comment]
Hi

Salesforce partner here.  We have implemented Give-WP (Salesforce and Recurring Giving plugins) for a few orgs now.  We are seeing an issue which is very closely related to this post.  Happy to start another fresh post, let me know.

Clients have reported this and we have replicated, in two different client environments:

1. Monthly donation subscribed to online.
2. Account/Contact/Recurring Donation (npe03__Recurring_Donation__c) all created fine
3. 1st donation appears in SF twice:  two opportunities. Both created at the same time, stage 'closed won'

One (DON1) is linked to the Recurring Donation (RD), the other (DON2) is not.
DON1 shows stage history: Pledged first, then Closed Won
DON2 shows stage history: only Closed Won

We have tried changing the NPSP 'installment opp auto creation' setting
From
Always create next installment
TO
Disable First Installment on Create

This time:
1. Monthly donation subscribed to online.
2. Account/Contact/Recurring Donation (npe03__Recurring_Donation__c) all created fine
3. One Opp created
DON1 not linked to RD
DON1 stage Pledged - no updates to the Stage

From previous posts, does GiveWP look for existing Opps first, deletes it if found and creates it's own Opp?

Based on the second test, something isn't quite right with the opp creation:
Stage should be Closed Won
Opp link to RD should be populated (Opportunity.npe03__Recurring_Donation__c)

Any insight into the main issue, or the second issue?  Anyone else seeing this?

Cheers

**Reply by jason-best** [reply]
SF Partner here as well. 
I've seen the exact same behavior. 
I ended up creating a custom function in WP to push over GiveWP values to the recurring Opps. I then have a flow that cleans up the extra Opp (DON2) in your first example above.

**Reply by merguk** [reply]
Awesome, thanks for confirming Jason :)

@jonwaldstein are you able to look into this?

**Reply by jonwaldstein** [reply]
@merguk yes we can look into this, although we'll need to get this into our feedback/support queue in order to prioritize. I'll go ahead and do that and report back with a link to monitor progress.

**Reply by jonwaldstein** [reply]
Here is the link: https://feedback.givewp.com/bug-reports/p/salesforce-and-recurring-donation-problem-with-custom-fields

**Reply by merguk** [reply]
@jason-best have you ever had to deal with historical recurring donations?  Do you have any insights to share?

https://github.com/impress-org/givewp/discussions/6871
